### ISSUE TITLE
Abend when mpd file and subfile shares the same name

### ISSUE THREAD

**Issue by trevorsandy** [leozide/leocad#533]
This sample file will cause an abend as a result of PieceInfo::CompareBoundingBox and lcModel::SubModelCompareBoundingBox triggering a runaway loop.

Steps:
1. Save this content to a _torso.mpd_ and copy the full path to clipboard `ctrl+c`
2. Launch LeoCAD and open the file dialogue `ctrl+o`
3. Navigate to and select _torso.mpd_ or enter file path in the _File name:_ dialogue box `ctrl+v` and click _Open_.

```
0 FILE torso.ldr
0 Torso
0 Name: torso.ldr
1 0 0 0 0 1 0 0 0 1 0 0 0 1 torso.ldr
0 NOFILE
0 FILE torso.ldr
0 Torso
0 Name: torso.ldr
1 14 0 0 0 0.985 0.12 0.12 -0.002 0.717 -0.697 -0.17 0.686 0.707 3820.dat
1 1 23.688 -26.76 9.884 1 0 0 0 1 0 0 0 1 973.dat
1 1 8.136 -17.76 9.884 0.985 -0.17 0 0.17 0.985 0 0 0 1 3818.dat
1 1 39.239 -17.76 9.884 0.985 0.17 0 -0.17 0.985 0 0 0 1 3819.dat
1 14 47.375 0 0 0.985 0.12 -0.12 -0.17 0.697 -0.697 0 0.707 0.707 3820.dat
0 NOFILE
```

Perhaps it could be good to handle this exception with an error message and gracefully unload the model file.

Cheers,

**Comment by trevorsandy** [comment]
With the latest build, the above sample file does not trigger an abend but it also does not render any display.

If I change the case of a letter in either the model or submodel file name, an abend is triggered on file load. 

For example, the example file below will trigger an abend:

```
0 FILE torso.ldr
0 Torso
0 Name: torso.ldr
1 0 0 0 0 1 0 0 0 1 0 0 0 1 toRso.ldr
0 NOFILE
0 FILE toRso.ldr
0 Torso
0 Name: toRso.ldr
1 14 0 0 0 0.985 0.12 0.12 -0.002 0.717 -0.697 -0.17 0.686 0.707 3820.dat
1 1 23.688 -26.76 9.884 1 0 0 0 1 0 0 0 1 973.dat
1 1 8.136 -17.76 9.884 0.985 -0.17 0 0.17 0.985 0 0 0 1 3818.dat
1 1 39.239 -17.76 9.884 0.985 0.17 0 -0.17 0.985 0 0 0 1 3819.dat
1 14 47.375 0 0 0.985 0.12 -0.12 -0.17 0.697 -0.697 0 0.707 0.707 3820.dat
0 NOFILE
```

Cheers,

**Comment by leozide** [comment]
I don't know if the first case where nothing draws is a bug. I see it as 2 submodels with the same name so the 2nd submodel is ignored and since the include line would cause an infinite recursion it's ignored as well.

I'll take a look at the new case but I think it should be handled the same way as the previous case and not draw anything because file names are case insensitive:

_Note: The file format specification states that filenames are case-insensitive. However, all official parts will be issued with lower-case only names._

I always thought the spec was a bit vague in this area so I'm not sure what's right here.

**Comment by rsbx** [comment]
I was recently looking for this in the SPECs and I was not able to find 
where it was specified one way or the other for type 1 lines. The only 
reference I found to case (in)sensitivity is in the 'LDraw File Format 
Specification' for the 'File Type' meta command. Where is says "In 
general, parsers should consider this line to be case-insensitive and 
free-format"; referring to the 'File Type' meta line. The Library and 
OMR specs are silent on the issue.

**Comment by trevorsandy** [comment]
@leozide, agreed. My suggestion was to trigger a message on encountering this configuration as it is almost certain a model file with this configuration would not have been intentionally defined to show an empty display.

Cheers,

**Comment by leozide** [comment]
Maybe auto rename is a better solution?

**Comment by trevorsandy** [comment]
Yep, that’s precisely what I do in LPub3D.

Cheers,

**Comment by rsbx** [comment]
Rename which one? FILE names in MPD files are de-facto 
case-insensitive[1] due to the history of the ambiguous spec and actual 
use. So, at this point it's pretty well determined that when this 
situation arises, there are significant issues with the MPD file and the 
correct(TM) is itself ambigious.

[1] And it's actually much worse: case-insensitivity is Unicode is 
complex and changes over time.

**Comment by trevorsandy** [comment]
For my scenario, I rename the top-level (model) file. 

I have seen and can think of only a narrow probability of encountering this use case. 

For me, it’s when a generated part (e.g. an LDCad generated PF Motor with cable) needs to be rendered as a part - i.e. an LPub3D PLI part. In such case, I create an mpd from the ‘part’ to pass to LeoCAD in order to capture the generated cable submodel. However, in creating the mpd I must choose to rename either the top-level name or submodel name, the submodel in this case is the part I’m rendering. As such, I rename the top-level name so the part/submodel name remains as it is used in the model file.

Cheers,

**Comment by trevorsandy** [comment]
> So, at this point it's pretty well determined that when this
situation arises, there are significant issues with the MPD file and the
correct(TM) is itself ambigious.

This statement is entirely dependent on the parser design and object model of the opening application. For example, the above sample models can be opened in LDView without issue.

Cheers,

**Comment by rsbx** [comment]
I would love it if the LDraw organization came down one way or the other 
on this issue; it would simplify things (over time as maintained tools 
converged on a common behavior) and make is easier for for users to 
diagnose the problem and apply a fix.

But at the moment, MPD files may work fine is some tools but not others. 
And this makes things frustrating for users. Silently masking the 
problem is rarely a real solution. At the very least, some kind of 
diagnostic should be provided to help the user understand why the tool 
appears to be broken (segfaults, renders something different than 
another tool, etc).
